torch/csrc/jit/passes/integer_value_refinement.cpp

```diff
@@ -0,0 +1,215 @@
+        c10::optional<int64_t> maybe_refine = tryFindRefinement(block_output);
+        if (!maybe_refine && other_block_refinements.count(block_output)) {
+          maybe_refine = other_block_refinements.at(block_output);
+        }
+        if (maybe_refine && *maybe_refine == *other_const_value) {
+          if_node->outputs().at(i)->replaceAllUsesWith(block_output);
+          changed_ = true;
+        }
+      }
+    }
+  }
+
+  // iteratively look through the block `b` for refinements or Value uses that
+  // can be refined, `block_refinements` are the refinements present starting at
+  // this block (and for all blocks dominated by this block).
+  IntegerRefinement RefineIntegerValues(
+      Block* b,
+      IntegerRefinement block_refinements) {
+    active_refinements_.push_back(&block_refinements);
+    for (Node* n : b->nodes()) {
+      if (n->matches("aten::eq(int a, int b) -> bool") ||
+          n->matches("aten::ne(int a, int b) -> bool")) {
+        for (size_t const_index : {0, 1}) {
+          if (auto ival = constant_as<int64_t>(n->inputs().at(const_index))) {
+            IntegerRefinement refine;
+            refine[n->inputs().at(1 - const_index)] = *ival;
+            info_[n->output()] = n->kind() == aten::eq
+                ? BooleanRefinementMapping::TrueRefinements(std::move(refine))
+                : BooleanRefinementMapping::FalseRefinements(std::move(refine));
+          }
+        }
+      }
+      for (size_t input = 0; input < n->inputs().size(); ++input) {
+        Value* input_v = n->inputs().at(input);
+        if (!input_v->type()->cast<IntType>()) {
+          continue;
+        }
+
+        if (auto refine = tryFindRefinement(input_v)) {
+          WithInsertPoint guard(n);
+          auto refine_constant =
+              graph_->insertConstant(static_cast<int64_t>(*refine));
+          n->replaceInputWith(input_v, refine_constant);
+          changed_ = true;
+        }
+      }
+
+      if (n->kind() == prim::If) {
+        IfView if_n(n);
+        bool has_cond_ref = info_.count(if_n.cond()) != 0;
+        IntegerRefinement empty;
+        auto true_block_refinements = RefineIntegerValues(
+            if_n.thenBlock(),
+            has_cond_ref ? info_[if_n.cond()].true_refine() : empty);
+        auto false_block_refinements = RefineIntegerValues(
+            if_n.elseBlock(),
+            has_cond_ref ? info_[if_n.cond()].false_refine() : empty);
+
+        removeIfNodeOutputsWithRefinements(
+            n, true_block_refinements, false_block_refinements);
+
+        joinIfRefinements(
+            n,
+            throwing_blocks_,
+            block_refinements,
+            true_block_refinements,
+            false_block_refinements,
+            info_);
+      } else {
+        handleCommonRefinentOperators(n, throwing_blocks_, info_);
+      }
+    }
+
+    // this is useful for things like if block outputs,
+    // where the output value node may not be defined in this block
```

Comment:
Nit: what does "output value node may not be defined" mean?

Reply:
Will clarify..

Reply:
To clarify a bit: I think the word 'node' simply needs to be deleted there :)